Source/WebCore/style/values/sizing/StyleAspectRatio.h

```diff
@@ -0,0 +1,118 @@
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "RenderStyleConstants.h"
+#include "StyleRatio.h"
+
+namespace WebCore {
+namespace Style {
+
+using namespace CSS::Literals;
+
+// <'aspect-ratio'> = auto || <ratio>
+// https://drafts.csswg.org/css-sizing-4/#propdef-aspect-ratio
+struct AspectRatio {
+    AspectRatio(CSS::Keyword::Auto)
+        : type { AspectRatioType::Auto }
+        , ratio { 0_css_number, 0_css_number }
+    {
+    }
+
+    AspectRatio(CSS::Keyword::Auto, Style::Ratio ratio)
+        : type { AspectRatioType::AutoAndRatio }
+        , ratio { ratio }
+    {
+    }
+
+    AspectRatio(Style::Ratio ratio)
+        : type { (!ratio.numerator.value || !ratio.denominator.value) ? AspectRatioType::AutoZero : AspectRatioType::Ratio }
+        , ratio { ratio }
+    {
+    }
+
+    bool isAuto() const             { return type == AspectRatioType::Auto; }
+    bool isAutoAndRatio() const     { return type == AspectRatioType::AutoAndRatio; }
+    bool isAutoZero() const         { return type == AspectRatioType::AutoZero; }
+    bool isRatio() const            { return type == AspectRatioType::Ratio; }
+
+    bool hasRatio() const           { return isRatio() || isAutoAndRatio(); }
+
+    std::optional<Style::Ratio> tryRatio() const
+    {
+        if (isAuto())
+            return { };
+        return ratio;
+    }
+
+    Number<CSS::Nonnegative> width() const { return ratio.numerator; }
+    Number<CSS::Nonnegative> height() const { return ratio.denominator; }
+
+    template<typename... F> decltype(auto) switchOn(F&&... f) const
+    {
+        auto visitor = WTF::makeVisitor(std::forward<F>(f)...);
+
+        switch (type) {
+        case AspectRatioType::Auto:
+            return visitor(CSS::Keyword::Auto { });
+        case AspectRatioType::AutoZero:
+        case AspectRatioType::Ratio:
+            return visitor(ratio);
+        case AspectRatioType::AutoAndRatio:
+            return visitor(SpaceSeparatedTuple { CSS::Keyword::Auto { }, ratio });
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    bool operator==(const AspectRatio&) const = default;
+
+private:
+    friend struct Blending<AspectRatio>;
+
+    AspectRatio(AspectRatioType type, Style::Ratio ratio)
+        : type { type }
+        , ratio { ratio }
+    {
+    }
+
+    AspectRatioType type;
+    Style::Ratio ratio;
```

Comment:
Don't we usually use the `m_` prefix on private data members? Maybe these should be public.